### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by SelmaNdapanda**
### In this pull request, @Kweeka1 and I implemented the following:  :arrow_down: 

- Set up the repository on GitHub and use Gitflow.  :heavy_check_mark: 
- Set up Devise for authentication.  :heavy_check_mark: 
- Set up RSpec and Capybara testing libraries.  :heavy_check_mark: 

#### We followed the layout of the wireframes provided. We however personalized the rest of the design including colors, typographies, spacings, etc.

- Login page and registration page:  :heavy_check_mark: 
    - Built with Devise.
- Food list:  :heavy_check_mark: 
    - Display a list of food added by the logged-in user as in the wireframe (for teams with only 2 people - display also quantity of a given food).
    - Lead to a form that allows users to add new food.
- Recipes list:  :heavy_check_mark: 
    - Display a list of recipes created by the logged-in user as in the wireframe.
    - Lead to recipe details.
    - If the user is the owner of the recipe, should allow the user to delete it.
- Public recipe list:  :heavy_check_mark: 
    - Display a list of all public recipes ordered by newest as in the wireframe.
    - Should lead to recipe details.
- Recipe details:  :heavy_check_mark: 
    - Display a toggle button that allows for a recipe to be made public or private.
    - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
    - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- General shopping list view:  :heavy_check_mark: 
    - Show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - Count the total food items and total price of the missing food.
- Made sure there are no N+1 queries happening.  :heavy_check_mark: 
- Created a navigation menu that allows users to open all of the pages you created.  :heavy_check_mark: 
- Wrote unit and integration tests  :heavy_check_mark:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,196 +47,88 @@ After you're finished please remove all the comments and instructions!
...
 
-This project is [MIT](./LICENSE) licensed.
+- Microverse 
+- Code Reviewers
 
-_NOTE: we recommend using the [MIT license](https://choosealicense.com/licenses/mit/) - you can set it up quickly by [using templates available on GitHub](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository). You can also use [any other license](https://choosealicense.com/licenses/) if you wish._
+## License 
 
-<p align="right">(<a href="#readme-top">back to top</a>)</p>
+This project is [MIT](./LICENSE) licensed.
```
Comment (danmainah) [inline README.md]: - [x] @SelmaNdapanda and @Kweeka1 kindly ensure your MIT license is working, its currently broken.

File: app/models/food.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  validates :measurement_unit, presence: true, length: { maximum: 20 }
+  validates :price, presence: true, numericality: { greater_than: 0 }
+  validates :quantity, presence: true, numericality: { greater_than: 0 }
+
+  # private
+
+  # def increment_total_food_items(recipe)
+  #   recipe.total_food_items.increment
+  #   recipe.save
+  # end
```
Comment (danmainah) [inline app/models/food.rb]: - [x] @team it's not good practice to have commented code in your codebase kindly delete it or uncomment it..

File: app/controllers/recipes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      render json: { status: 'Food item successfully added' }, status: 201
+    else
+      render :ingredients, status: :unprocessable_entity
+    end
+  end
+
+  def remove_ingredient
+    recipe_food = RecipeFood.find_by(id: recipe_params[:recipe_food_id])
+    recipe_food.destroy
+    redirect_to recipe_path(recipe_params[:recipe_id])
```
Comment (danmainah) [inline app/controllers/recipes_controller.rb]: - [x] @team its good practice to have both success and failure paths for methods that have  create, delete, or update  actions, kindly ensure your methods have both paths.

File: app/controllers/recipes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      redirect_to recipe_path @recipe
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    recipe = Recipe.find_by(id: recipe_params[:recipe_id])
+    recipe.destroy
+    redirect_to recipes_path
```
Comment (danmainah) [inline app/controllers/recipes_controller.rb]: - [x] @team its good practice to have both success and failure paths for methods that have  create, delete, or update  actions, kindly ensure your methods have both paths.

File: app/controllers/foods_controller.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    else
+      render :new, alert: 'Cannot create food, retry again.'
+    end
+  end
+
+  def destroy
+    @food = Food.find(params[:id])
+    @food.destroy
+
+    redirect_to foods_path, notice: 'Food was successfully destroyed.'
```
Comment (danmainah) [inline app/controllers/foods_controller.rb]: - [x] @team its good practice to have both success and failure paths for methods that have  create, delete, or update  actions, kindly ensure your methods have both paths.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @SelmaNdapanda and @Kweeka1 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️ No linter errors
✔️ Gitflow was correctly used
✔️  Good design


# Required Changes ♻️

- [ ] @SelmaNdapanda and @Kweeka1 kindly add the shopping list page, which shows a list of food that is missing for all recipes of the logged-in user and counts the total food items and total price of the missing food. Also, ensure to add a link in the navigation bar for the page.
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: README.md (hunk 1 above)
Comment (Abenezer-Tilahun) [reply README.md]: Fixed.

File: app/models/food.rb (hunk 2 above)
Comment (Abenezer-Tilahun) [reply app/models/food.rb]: Fixed.

File: app/controllers/recipes_controller.rb (hunk 3 above)
Comment (Abenezer-Tilahun) [reply app/controllers/recipes_controller.rb]: Fixed.

File: app/controllers/recipes_controller.rb (hunk 4 above)
Comment (Abenezer-Tilahun) [reply app/controllers/recipes_controller.rb]: Fixed.

File: app/controllers/foods_controller.rb (hunk 5 above)
Comment (Abenezer-Tilahun) [reply app/controllers/foods_controller.rb]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @SelmaNdapanda @Kweeka1 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**